### PR TITLE
Refactor, add type hints to get_chunk() in datatypes

### PR DESCRIPTION
**Pull request by jdavcs**
This modification was triggered by a mypy error (incompatible method signature with parent class). This was part of a larger upcoming PR (adding type hints to datatypes), but since the method had to be modified significantly, I think it makes sense to have this as a standalone PR.

**Details**
The are 2 variations of the `get_chunk` method signature on tabular datatypes: one taking a chunk index, the other one taking an absolute offset. The method is [called from the legacy dataset controller](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/controllers/dataset.py#L574); the argument could be interpreted either as an absolute offset or a chunk index. This has not been a problem because the only call to this method had `0` hardcoded as the argument, so it didn't matter how it was interpreted (if it were to be called with an argument `> 0`, it would have become a bug)

**Note**
The original implementation used a chunk index. As a result, prior to reading in the next chunk (if not at the beginning of the file), it skipped over characters until the next newline character. Thus, if the calculated offset pointed to the middle of a line, that partial line was not read. This is no longer needed since we rely on the client to provide the absolute offset. (implemented by @dannon [here](https://github.com/galaxyproject/galaxy/commit/84d152082009418f2fd5dfe0e9001e8924a678fc))

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by jdavcs** [comment]
Test failures unrelated.

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -571,16 +577,16 @@ def groom_dataset_content(self, file_name):
         # Remove temp file and empty temporary directory
         os.rmdir(tmp_dir)
 
-    def get_chunk(self, trans, dataset, offset=0, ck_size=None):
+    def get_chunk(self, trans, dataset: "DatasetInstance", offset: int = 0, ck_size: Optional[int] = None) -> str:
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py:580]: This method feels so wrong in many ways (the json dump, the pysam use, the fact that you should probably address files like this through byte range requests) ... I guess it doesn't hurt to do type annotations, but I'd rather delete it ;)

File: lib/galaxy/datatypes/tabular.py [hunk 2]
```diff
@@ -107,7 +113,17 @@ def displayable(self, dataset: DatasetInstance):
         except Exception:
             return False
 
-    def get_chunk(self, trans, dataset, offset=0, ck_size=None):
+    def get_chunk(self, trans, dataset: "DatasetInstance", offset: int = 0, ck_size: Optional[int] = None) -> str:
+        ck_data, last_read = self._read_chunk(trans, dataset, offset, ck_size)
+        return dumps(
+            {
+                "ck_data": util.unicodify(ck_data),
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/tabular.py:120]: The default mode is "r" for get_fileobj, so the unicodify shouldn't be necessary (not a new thing, no need to fix it here)

**Review by mvdbeek — APPROVED** [review]

File: lib/galaxy/datatypes/binary.py (hunk 1 above)
Comment (jdavcs) [reply lib/galaxy/datatypes/binary.py:580]: Yeah, I'd be happy to delete it, but I don't understand the its usage well enough (called from a legacy controller + from the `display_data` method). I'm adding this to my "to understand/to fix" long term todo list :)

File: lib/galaxy/datatypes/tabular.py (hunk 2 above)
Comment (jdavcs) [reply lib/galaxy/datatypes/tabular.py:120]: Thanks, leaving it as is for now.

**Comment by jdavcs** [comment]
Thank you for the review, @mvdbeek!

**Pull request merged**